# Importação da BigQuery


In [33]:
!pip install --upgrade google-cloud-bigquery --quiet

In [34]:
from google.colab import auth
auth.authenticate_user()

In [35]:
def consulta():
    import re
    from google.cloud import bigquery

    # Configurações fixas
    project_id = "guilherme-454214"
    dataset = "locadora"

    # Entrada do usuário
    sql = input("Digite sua query (sem project_id.dataset):\n")
    nome_df = input("Informe o nome para o DataFrame:\n")

    # Ajusta FROM e JOINs com project.dataset
    sql = re.sub(r"FROM\s+(\w+)", lambda m: f"FROM `{project_id}.{dataset}.{m.group(1)}`", sql, flags=re.I)
    sql = re.sub(r"(JOIN|LEFT JOIN|RIGHT JOIN|INNER JOIN|FULL JOIN|OUTER JOIN)\s+(\w+)",
                 lambda m: f"{m.group(1)} `{project_id}.{dataset}.{m.group(2)}`", sql, flags=re.I)
    print(sql)

    # Executa no BigQuery
    df = bigquery.Client(project=project_id).query(sql).to_dataframe()
    globals()[nome_df] = df

    print(f"\n✅ DataFrame '{nome_df}' criado com sucesso! ({len(df)} linhas)")
    display(df)

# Exercícios SQL

Este notebook contém consultas SQL para responder aos exercícios.

## Exercício 1

Selecione todos os dados da tabela cliente.

```sql
-- Escreva sua consulta aqui
```

In [4]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT * FROM cliente
Informe o nome para o DataFrame:
Q1
SELECT * FROM `guilherme-454214.locadora.cliente`

✅ DataFrame 'Q1' criado com sucesso! (5 linhas)


codcliente              nome           cidade sexo estado estadocivil
0           1         Ana Silva  Duque de Caxias    F     RJ           C
1           2     Bruna Pereira          Niterói    F     RJ           C
2           3  Túlio Nascimento  Duque de Caxias    M     RJ           S
3           4    Fernando Souza         Campinas    M     SP           S
4           5     Lúcia Andrade        São Paulo    F     SP           C

## Exercício 2

Liste todos os clientes que moram na cidade de "São Paulo".

```sql
-- Escreva sua consulta aqui
```

In [8]:
consulta()



Digite sua query (sem project_id.dataset):
SELECT * FROM cliente WHERE cidade = 'São Paulo'
Informe o nome para o DataFrame:
Q2
SELECT * FROM `guilherme-454214.locadora.cliente` WHERE cidade = 'São Paulo'

✅ DataFrame 'Q2' criado com sucesso! (1 linhas)


codcliente           nome     cidade sexo estado estadocivil
0           5  Lúcia Andrade  São Paulo    F     SP           C

## Exercício 3

Liste os estados diferentes presentes na tabela cliente.

```sql
-- Escreva sua consulta aqui
```

In [9]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT DISTINCT estado FROM cliente
Informe o nome para o DataFrame:
Q3
SELECT DISTINCT estado FROM `guilherme-454214.locadora.cliente`

✅ DataFrame 'Q3' criado com sucesso! (2 linhas)


estado
0     RJ
1     SP

## Exercício 4

Liste todos os clientes ordenados pelo nome em ordem alfabética.

```sql
-- Escreva sua consulta aqui
```

In [10]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT * FROM cliente ORDER BY nome
Informe o nome para o DataFrame:
Q4
SELECT * FROM `guilherme-454214.locadora.cliente` ORDER BY nome

✅ DataFrame 'Q4' criado com sucesso! (5 linhas)


codcliente              nome           cidade sexo estado estadocivil
0           1         Ana Silva  Duque de Caxias    F     RJ           C
1           2     Bruna Pereira          Niterói    F     RJ           C
2           4    Fernando Souza         Campinas    M     SP           S
3           5     Lúcia Andrade        São Paulo    F     SP           C
4           3  Túlio Nascimento  Duque de Caxias    M     RJ           S

## Exercício 5

Encontre todos os clientes que são do sexo feminino e solteiros.

```sql
-- Escreva sua consulta aqui
```

In [11]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT * FROM cliente WHERE sexo = 'F' AND estadocivil = 'S'
Informe o nome para o DataFrame:
Q5
SELECT * FROM `guilherme-454214.locadora.cliente` WHERE sexo = 'F' AND estadocivil = 'S'

✅ DataFrame 'Q5' criado com sucesso! (0 linhas)


Empty DataFrame
Columns: [codcliente, nome, cidade, sexo, estado, estadocivil]
Index: []

## Exercício 6

Liste todos os clientes cujo nome começa com a letra 'A'.

```sql
-- Escreva sua consulta aqui
```

In [12]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT * FROM cliente WHERE nome LIKE 'A%'
Informe o nome para o DataFrame:
Q6
SELECT * FROM `guilherme-454214.locadora.cliente` WHERE nome LIKE 'A%'

✅ DataFrame 'Q6' criado com sucesso! (1 linhas)


codcliente       nome           cidade sexo estado estadocivil
0           1  Ana Silva  Duque de Caxias    F     RJ           C

## Exercício 7

Selecione os 3 primeiros registros da tabela cliente.

```sql
-- Escreva sua consulta aqui
```

In [13]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT * FROM cliente LIMIT 3
Informe o nome para o DataFrame:
Q7
SELECT * FROM `guilherme-454214.locadora.cliente` LIMIT 3

✅ DataFrame 'Q7' criado com sucesso! (3 linhas)


codcliente              nome           cidade sexo estado estadocivil
0           1         Ana Silva  Duque de Caxias    F     RJ           C
1           2     Bruna Pereira          Niterói    F     RJ           C
2           3  Túlio Nascimento  Duque de Caxias    M     RJ           S

## Exercício 8

Liste todos os carros que são da marca 'Ford' ou 'Fiat'.

```sql
-- Escreva sua consulta aqui
```

In [17]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT ca.*, ma.marca FROM carro ca JOIN marca ma ON ca.codmarca = ma.codmarca WHERE ma.marca IN ('Ford', 'Fiat')
Informe o nome para o DataFrame:
Q8
SELECT ca.*, ma.marca FROM `guilherme-454214.locadora.carro` ca JOIN `guilherme-454214.locadora.marca` ma ON ca.codmarca = ma.codmarca WHERE ma.marca IN ('Ford', 'Fiat')

✅ DataFrame 'Q8' criado com sucesso! (2 linhas)


codcarro  codmarca modelo  valor marca
0         1         1     Ka    100  Ford
1         2         2   Argo    150  Fiat

## Exercício 9

Encontre todos os carros cujo valor está entre 100 e 150.

```sql
-- Escreva sua consulta aqui
```

In [18]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT * FROM carro WHERE valor BETWEEN 100 AND 150
Informe o nome para o DataFrame:
Q9
SELECT * FROM `guilherme-454214.locadora.carro` WHERE valor BETWEEN 100 AND 150

✅ DataFrame 'Q9' criado com sucesso! (4 linhas)


codcarro  codmarca modelo  valor
0         1         1     Ka    100
1         5         5   Kwid    120
2         2         2   Argo    150
3         4         4   Polo    150

## Exercício 10

Conte quantos aluguéis foram realizados.

```sql
-- Escreva sua consulta aqui
```

In [19]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT COUNT(*) AS total_alugueis FROM aluguel 
Informe o nome para o DataFrame:
Q10
SELECT COUNT(*) AS total_alugueis FROM `guilherme-454214.locadora.aluguel` 

✅ DataFrame 'Q10' criado com sucesso! (1 linhas)


total_alugueis
0              10

## Exercício 11

Calcule o valor total de todos os carros disponíveis para aluguel.

```sql
-- Escreva sua consulta aqui
```

In [20]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT SUM(valor) AS valor_total FROM carro
Informe o nome para o DataFrame:
Q11
SELECT SUM(valor) AS valor_total FROM `guilherme-454214.locadora.carro`

✅ DataFrame 'Q11' criado com sucesso! (1 linhas)


valor_total
0          690

## Exercício 12

Calcule o valor médio dos carros.

```sql
-- Escreva sua consulta aqui
```

In [21]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT AVG(valor) AS valor_medio FROM carro
Informe o nome para o DataFrame:
Q12
SELECT AVG(valor) AS valor_medio FROM `guilherme-454214.locadora.carro`

✅ DataFrame 'Q12' criado com sucesso! (1 linhas)


valor_medio
0        138.0

## Exercício 13

Encontre o menor e o maior valor de aluguel de carros.

```sql
-- Escreva sua consulta aqui
```

In [22]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT MIN(valor) AS menor_valor, MAX(valor) AS maior_valor FROM carro
Informe o nome para o DataFrame:
Q13
SELECT MIN(valor) AS menor_valor, MAX(valor) AS maior_valor FROM `guilherme-454214.locadora.carro`

✅ DataFrame 'Q13' criado com sucesso! (1 linhas)


menor_valor  maior_valor
0          100          170

## Exercício 14

Agrupe os clientes por estado e conte quantos clientes existem em cada estado.

```sql
-- Escreva sua consulta aqui
```

In [23]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT estado, COUNT(*) AS total_clientes FROM cliente GROUP BY estado
Informe o nome para o DataFrame:
Q14
SELECT estado, COUNT(*) AS total_clientes FROM `guilherme-454214.locadora.cliente` GROUP BY estado

✅ DataFrame 'Q14' criado com sucesso! (2 linhas)


estado  total_clientes
0     RJ               3
1     SP               2

## Exercício 15

Agrupe os aluguéis por cliente e filtre apenas aqueles que fizeram mais de 2 aluguéis.

```sql
-- Escreva sua consulta aqui
```

In [24]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT codcliente, COUNT(*) AS total_alugueis FROM aluguel GROUP BY codcliente HAVING total_alugueis > 2
Informe o nome para o DataFrame:
Q15
SELECT codcliente, COUNT(*) AS total_alugueis FROM `guilherme-454214.locadora.aluguel` GROUP BY codcliente HAVING total_alugueis > 2

✅ DataFrame 'Q15' criado com sucesso! (2 linhas)


codcliente  total_alugueis
0           2               3
1           1               3

## Exercício 16

Liste todos os aluguéis, mostrando o nome do cliente e o modelo do carro alugado.

```sql
-- Escreva sua consulta aqui
```

In [25]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT cli.nome, ca.modelo FROM aluguel a JOIN cliente cli ON a.codcliente = cli.codcliente JOIN carro ca ON a.codcarro = ca.codcarro
Informe o nome para o DataFrame:
Q16
SELECT cli.nome, ca.modelo FROM `guilherme-454214.locadora.aluguel` a JOIN `guilherme-454214.locadora.cliente` cli ON a.codcliente = cli.codcliente JOIN `guilherme-454214.locadora.carro` ca ON a.codcarro = ca.codcarro

✅ DataFrame 'Q16' criado com sucesso! (10 linhas)


nome modelo
0     Bruna Pereira     Ka
1     Bruna Pereira     Ka
2         Ana Silva     Ka
3  Túlio Nascimento     Ka
4  Túlio Nascimento   Argo
5     Lúcia Andrade   Argo
6     Lúcia Andrade   Argo
7     Bruna Pereira   Onix
8         Ana Silva   Polo
9         Ana Silva   Polo

## Exercício 17

Liste todos os carros e, se houver, as informações sobre os aluguéis realizados desses carros.

```sql
-- Escreva sua consulta aqui
```

In [27]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT ca.*, a.* FROM carro ca LEFT JOIN aluguel a ON ca.codcarro = a.codcarro
Informe o nome para o DataFrame:
Q17
SELECT ca.*, a.* FROM `guilherme-454214.locadora.carro` ca LEFT JOIN `guilherme-454214.locadora.aluguel` a ON ca.codcarro = a.codcarro

✅ DataFrame 'Q17' criado com sucesso! (11 linhas)


codcarro  codmarca modelo  valor  codaluguel  codcliente  codcarro_1  \
0          1         1     Ka    100           2           2           1   
1          1         1     Ka    100           3           2           1   
2          1         1     Ka    100           7           1           1   
3          1         1     Ka    100          10           3           1   
4          5         5   Kwid    120        <NA>        <NA>        <NA>   
5          2         2   Argo    150           1           3           2   
6          2         2   Argo    150           8           5           2   
7          2         2   Argo    150           9           5           2   
8          4         4   Polo    150           5           1           4   
9          4         4   Polo    150           6           1           4   
10         3         3   Onix    170           4           2           3   

   data_aluguel  
0    2023-04-02  
1    2023-04-03  
2    2023-04-15  
3    2023-04-25  
4           NaT  
5    2023-04-01  
6    2023-04-19  
7    2023-04-21  
8    2023-04-05  
9    2023-04-13  
10   2023-04-04

## Exercício 18

Liste todos os aluguéis e as informações sobre os carros alugados, mesmo que o carro não esteja mais no banco de dados.

```sql
-- Escreva sua consulta aqui
```

In [28]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT a.*, ca.* FROm aluguel a LEFT JOIN carro ca ON a.codcarro = ca.codcarro
Informe o nome para o DataFrame:
Q18
SELECT a.*, ca.* FROM `guilherme-454214.locadora.aluguel` a LEFT JOIN `guilherme-454214.locadora.carro` ca ON a.codcarro = ca.codcarro

✅ DataFrame 'Q18' criado com sucesso! (10 linhas)


codaluguel  codcliente  codcarro data_aluguel  codcarro_1  codmarca modelo  \
0           2           2         1   2023-04-02           1         1     Ka   
1           3           2         1   2023-04-03           1         1     Ka   
2           7           1         1   2023-04-15           1         1     Ka   
3          10           3         1   2023-04-25           1         1     Ka   
4           1           3         2   2023-04-01           2         2   Argo   
5           8           5         2   2023-04-19           2         2   Argo   
6           9           5         2   2023-04-21           2         2   Argo   
7           4           2         3   2023-04-04           3         3   Onix   
8           5           1         4   2023-04-05           4         4   Polo   
9           6           1         4   2023-04-13           4         4   Polo   

   valor  
0    100  
1    100  
2    100  
3    100  
4    150  
5    150  
6    150  
7    170  
8    150  
9    150

## Exercício 19

Liste todos os clientes e os carros que alugaram, incluindo clientes que não alugaram e carros que não foram alugados.

```sql
-- Escreva sua consulta aqui
```

In [38]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT cli.nome, ca.modelo  FROM cliente cli LEFT JOIN aluguel a ON cli.codcliente = a.codcliente LEFT JOIN carro ca ON a.codcarro = ca.codcarro UNION ALL SELECT cli.nome, ca.modelo  FROM carro ca LEFT JOIN aluguel a ON ca.codcarro = a.codcarro LEFT JOIN cliente cli ON a.codcliente = cli.codcliente
Informe o nome para o DataFrame:
Q19
SELECT cli.nome, ca.modelo  FROM `guilherme-454214.locadora.cliente` cli LEFT JOIN `guilherme-454214.locadora.aluguel` a ON cli.codcliente = a.codcliente LEFT JOIN `guilherme-454214.locadora.carro` ca ON a.codcarro = ca.codcarro UNION ALL SELECT cli.nome, ca.modelo  FROM `guilherme-454214.locadora.carro` ca LEFT JOIN `guilherme-454214.locadora.aluguel` a ON ca.codcarro = a.codcarro LEFT JOIN `guilherme-454214.locadora.cliente` cli ON a.codcliente = cli.codcliente

✅ DataFrame 'Q19' criado com sucesso! (22 linhas)


nome modelo
0      Bruna Pereira     Ka
1      Bruna Pereira     Ka
2          Ana Silva     Ka
3   Túlio Nascimento     Ka
4               None   Kwid
5   Túlio Nascimento   Argo
6      Lúcia Andrade   Argo
7      Lúcia Andrade   Argo
8          Ana Silva   Polo
9          Ana Silva   Polo
10     Bruna Pereira   Onix
11         Ana Silva     Ka
12         Ana Silva   Polo
13         Ana Silva   Polo
14     Bruna Pereira     Ka
15     Bruna Pereira     Ka
16     Bruna Pereira   Onix
17  Túlio Nascimento     Ka
18  Túlio Nascimento   Argo
19    Fernando Souza   None
20     Lúcia Andrade   Argo
21     Lúcia Andrade   Argo

## Exercício 20

Combine os resultados de duas consultas: uma que lista clientes solteiros e outra que lista clientes casados.

```sql
-- Escreva sua consulta aqui
```

In [37]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT * FROM cliente WHERE estadocivil = 'S' UNION ALL SELECT * FROM cliente WHERE estadocivil = 'C'
Informe o nome para o DataFrame:
Q20
SELECT * FROM `guilherme-454214.locadora.cliente` WHERE estadocivil = 'S' UNION ALL SELECT * FROM `guilherme-454214.locadora.cliente` WHERE estadocivil = 'C'

✅ DataFrame 'Q20' criado com sucesso! (5 linhas)


codcliente              nome           cidade sexo estado estadocivil
0           3  Túlio Nascimento  Duque de Caxias    M     RJ           S
1           4    Fernando Souza         Campinas    M     SP           S
2           1         Ana Silva  Duque de Caxias    F     RJ           C
3           2     Bruna Pereira          Niterói    F     RJ           C
4           5     Lúcia Andrade        São Paulo    F     SP           C

## Exercício 21

Encontre os clientes que alugaram o carro com modelo 'Onix'.

```sql
-- Escreva sua consulta aqui
```

In [39]:
consulta()


Digite sua query (sem project_id.dataset):
SELECT cli.nome FROM cliente cli JOIN aluguel a ON cli.codcliente = a.codcliente JOIN carro ca ON a.codcarro = ca.codcarro WHERE ca.modelo = 'Onix'
Informe o nome para o DataFrame:
Q21
SELECT cli.nome FROM `guilherme-454214.locadora.cliente` cli JOIN `guilherme-454214.locadora.aluguel` a ON cli.codcliente = a.codcliente JOIN `guilherme-454214.locadora.carro` ca ON a.codcarro = ca.codcarro WHERE ca.modelo = 'Onix'

✅ DataFrame 'Q21' criado com sucesso! (1 linhas)


nome
0  Bruna Pereira

## Exercício 22

Liste os modelos de carros e o número de vezes que cada um foi alugado.

```sql
-- Escreva sua consulta aqui
```

In [40]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT ca.modelo, COUNT(a.codaluguel) AS total_alugueis FROM carro ca LEFT JOIN aluguel a ON ca.codcarro = a.codcarro GROUP BY ca.modelo
Informe o nome para o DataFrame:
Q22
SELECT ca.modelo, COUNT(a.codaluguel) AS total_alugueis FROM `guilherme-454214.locadora.carro` ca LEFT JOIN `guilherme-454214.locadora.aluguel` a ON ca.codcarro = a.codcarro GROUP BY ca.modelo

✅ DataFrame 'Q22' criado com sucesso! (5 linhas)


modelo  total_alugueis
0     Ka               4
1   Kwid               0
2   Argo               3
3   Polo               2
4   Onix               1

## Exercício 23

Liste o nome dos clientes e o número de aluguéis que cada um fez.

```sql
-- Escreva sua consulta aqui
```

In [44]:
consulta()


Digite sua query (sem project_id.dataset):
SELECT cli.nome, COUNT(a.codaluguel) AS total_alugueis  FROM cliente cli LEFT JOIN aluguel a ON cli.codcliente = a.codcliente GROUP BY cli.codcliente, cli.nome
Informe o nome para o DataFrame:
Q23
SELECT cli.nome, COUNT(a.codaluguel) AS total_alugueis  FROM `guilherme-454214.locadora.cliente` cli LEFT JOIN `guilherme-454214.locadora.aluguel` a ON cli.codcliente = a.codcliente GROUP BY cli.codcliente, cli.nome

✅ DataFrame 'Q23' criado com sucesso! (5 linhas)


nome  total_alugueis
0         Ana Silva               3
1     Bruna Pereira               3
2  Túlio Nascimento               2
3    Fernando Souza               0
4     Lúcia Andrade               2

## Exercício 24

Crie uma consulta que retorne o nome dos clientes e uma coluna adicional chamada status_civil que classifica os clientes como "Solteiro" ou "Casado".

```sql
-- Escreva sua consulta aqui
```

In [43]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT nome, CASE estadocivil WHEN 'S' THEN 'Solteiro' WHEN 'C' THEN 'Casado' END AS status_civil FROM cliente
Informe o nome para o DataFrame:
Q24
SELECT nome, CASE estadocivil WHEN 'S' THEN 'Solteiro' WHEN 'C' THEN 'Casado' END AS status_civil FROM `guilherme-454214.locadora.cliente`

✅ DataFrame 'Q24' criado com sucesso! (5 linhas)


nome status_civil
0         Ana Silva       Casado
1     Bruna Pereira       Casado
2  Túlio Nascimento     Solteiro
3    Fernando Souza     Solteiro
4     Lúcia Andrade       Casado

## Exercício 25

Liste os nomes dos clientes e as datas em que alugaram carros.

```sql
-- Escreva sua consulta aqui
```

In [45]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT cli.nome, a.data_aluguel FROM cliente cli JOIN aluguel a ON cli.codcliente = a.codcliente
Informe o nome para o DataFrame:
Q25
SELECT cli.nome, a.data_aluguel FROM `guilherme-454214.locadora.cliente` cli JOIN `guilherme-454214.locadora.aluguel` a ON cli.codcliente = a.codcliente

✅ DataFrame 'Q25' criado com sucesso! (10 linhas)


nome data_aluguel
0         Ana Silva   2023-04-15
1         Ana Silva   2023-04-05
2         Ana Silva   2023-04-13
3     Bruna Pereira   2023-04-02
4     Bruna Pereira   2023-04-03
5     Bruna Pereira   2023-04-04
6  Túlio Nascimento   2023-04-25
7  Túlio Nascimento   2023-04-01
8     Lúcia Andrade   2023-04-19
9     Lúcia Andrade   2023-04-21

## Exercício 26

Liste todos os modelos de carros e suas respectivas marcas.

```sql
-- Escreva sua consulta aqui
```

In [46]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT ca.modelo, ma.marca FROM carro ca JOIN marca ma ON ca.codmarca = ma.codmarca
Informe o nome para o DataFrame:
Q26
SELECT ca.modelo, ma.marca FROM `guilherme-454214.locadora.carro` ca JOIN `guilherme-454214.locadora.marca` ma ON ca.codmarca = ma.codmarca

✅ DataFrame 'Q26' criado com sucesso! (5 linhas)


modelo       marca
0     Ka        Ford
1   Kwid     Renault
2   Argo        Fiat
3   Polo  Volkswagen
4   Onix   Chevrolet

## Exercício 27

Liste todos os aluguéis e os nomes dos clientes que os realizaram.

```sql
-- Escreva sua consulta aqui
```

In [49]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT a.*, cli.nome FROM aluguel a JOIN cliente cli ON a.codcliente = cli.codcliente
Informe o nome para o DataFrame:
Q27
SELECT a.*, cli.nome FROM `guilherme-454214.locadora.aluguel` a JOIN `guilherme-454214.locadora.cliente` cli ON a.codcliente = cli.codcliente

✅ DataFrame 'Q27' criado com sucesso! (10 linhas)


codaluguel  codcliente  codcarro data_aluguel              nome
0           2           2         1   2023-04-02     Bruna Pereira
1           3           2         1   2023-04-03     Bruna Pereira
2           7           1         1   2023-04-15         Ana Silva
3          10           3         1   2023-04-25  Túlio Nascimento
4           1           3         2   2023-04-01  Túlio Nascimento
5           8           5         2   2023-04-19     Lúcia Andrade
6           9           5         2   2023-04-21     Lúcia Andrade
7           4           2         3   2023-04-04     Bruna Pereira
8           5           1         4   2023-04-05         Ana Silva
9           6           1         4   2023-04-13         Ana Silva

## Exercício 28

Encontre o nome dos clientes e os modelos de carros que alugaram.

```sql
-- Escreva sua consulta aqui
```

In [50]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT cli.nome, ca.modelo FROM cliente cli JOIN aluguel a ON cli.codcliente = a.codcliente JOIN carro ca ON a.codcarro = ca.codcarro
Informe o nome para o DataFrame:
Q28
SELECT cli.nome, ca.modelo FROM `guilherme-454214.locadora.cliente` cli JOIN `guilherme-454214.locadora.aluguel` a ON cli.codcliente = a.codcliente JOIN `guilherme-454214.locadora.carro` ca ON a.codcarro = ca.codcarro

✅ DataFrame 'Q28' criado com sucesso! (10 linhas)


nome modelo
0         Ana Silva     Ka
1         Ana Silva   Polo
2         Ana Silva   Polo
3     Bruna Pereira     Ka
4     Bruna Pereira     Ka
5     Bruna Pereira   Onix
6  Túlio Nascimento     Ka
7  Túlio Nascimento   Argo
8     Lúcia Andrade   Argo
9     Lúcia Andrade   Argo

## Exercício 29

Liste todos os carros e, se houver, mostre as datas em que foram alugados.

```sql
-- Escreva sua consulta aqui
```

In [51]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT ca.*, a.data_aluguel FROM carro ca LEFT JOIN aluguel a ON ca.codcarro = a.codcarro
Informe o nome para o DataFrame:
Q29
SELECT ca.*, a.data_aluguel FROM `guilherme-454214.locadora.carro` ca LEFT JOIN `guilherme-454214.locadora.aluguel` a ON ca.codcarro = a.codcarro

✅ DataFrame 'Q29' criado com sucesso! (11 linhas)


codcarro  codmarca modelo  valor data_aluguel
0          1         1     Ka    100   2023-04-02
1          1         1     Ka    100   2023-04-03
2          1         1     Ka    100   2023-04-15
3          1         1     Ka    100   2023-04-25
4          5         5   Kwid    120          NaT
5          2         2   Argo    150   2023-04-01
6          2         2   Argo    150   2023-04-19
7          2         2   Argo    150   2023-04-21
8          4         4   Polo    150   2023-04-05
9          4         4   Polo    150   2023-04-13
10         3         3   Onix    170   2023-04-04

## Exercício 30

Encontre os clientes que moram na mesma cidade e estado e que alugaram carros.

```sql
-- Escreva sua consulta aqui
```

In [52]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT c1.nome AS cliente1, c2.nome AS cliente2, c1.cidade, c1.estado FROM cliente c1 JOIN cliente c2 ON c1.cidade = c2.cidade AND c1.estado = c2.estado AND c1.codcliente < c2.codcliente WHERE EXISTS (SELECT 1 FROM aluguel a WHERE a.codcliente = c1.codcliente) AND EXISTS (SELECT 1 FROM aluguel a WHERE a.codcliente = c2.codcliente)
Informe o nome para o DataFrame:
Q30
SELECT c1.nome AS cliente1, c2.nome AS cliente2, c1.cidade, c1.estado FROM `guilherme-454214.locadora.cliente` c1 JOIN `guilherme-454214.locadora.cliente` c2 ON c1.cidade = c2.cidade AND c1.estado = c2.estado AND c1.codcliente < c2.codcliente WHERE EXISTS (SELECT 1 FROM `guilherme-454214.locadora.aluguel` a WHERE a.codcliente = c1.codcliente) AND EXISTS (SELECT 1 FROM `guilherme-454214.locadora.aluguel` a WHERE a.codcliente = c2.codcliente)

✅ DataFrame 'Q30' criado com sucesso! (1 linhas)


cliente1          cliente2           cidade estado
0  Ana Silva  Túlio Nascimento  Duque de Caxias     RJ

## Exercício 31

Liste os nomes dos clientes e os carros que eles alugaram, incluindo os clientes que não alugaram nenhum carro.

```sql
-- Escreva sua consulta aqui
```

In [53]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT cli.nome, ca.modelo FROM cliente cli LEFT JOIN aluguel a ON cli.codcliente = a.codcliente LEFT JOIN carro ca ON a.codcarro = ca.codcarro
Informe o nome para o DataFrame:
Q31
SELECT cli.nome, ca.modelo FROM `guilherme-454214.locadora.cliente` cli LEFT JOIN `guilherme-454214.locadora.aluguel` a ON cli.codcliente = a.codcliente LEFT JOIN `guilherme-454214.locadora.carro` ca ON a.codcarro = ca.codcarro

✅ DataFrame 'Q31' criado com sucesso! (11 linhas)


nome modelo
0          Ana Silva     Ka
1          Ana Silva   Polo
2          Ana Silva   Polo
3      Bruna Pereira     Ka
4      Bruna Pereira     Ka
5      Bruna Pereira   Onix
6   Túlio Nascimento     Ka
7   Túlio Nascimento   Argo
8     Fernando Souza   None
9      Lúcia Andrade   Argo
10     Lúcia Andrade   Argo

## Exercício 32

Liste todas as marcas e, se houver, os modelos de carros associados.

```sql
-- Escreva sua consulta aqui
```

In [56]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT ma.marca, STRING_AGG(ca.modelo, ',') AS modelos  FROM marca ma LEFT JOIN carro ca ON ma.codmarca = ca.codmarca GROUP BY ma.marca;
Informe o nome para o DataFrame:
Q32
SELECT ma.marca, STRING_AGG(ca.modelo, ',') AS modelos  FROM `guilherme-454214.locadora.marca` ma LEFT JOIN `guilherme-454214.locadora.carro` ca ON ma.codmarca = ca.codmarca GROUP BY ma.marca;

✅ DataFrame 'Q32' criado com sucesso! (5 linhas)


marca modelos
0        Ford      Ka
1        Fiat    Argo
2   Chevrolet    Onix
3  Volkswagen    Polo
4     Renault    Kwid

## Exercício 33

Encontre todos os aluguéis e o estado civil dos clientes que alugaram os carros.

```sql
-- Escreva sua consulta aqui
```

In [55]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT a.*, cli.estadocivil FROM aluguel a JOIN cliente cli ON a.codcliente = cli.codcliente
Informe o nome para o DataFrame:
Q33
SELECT a.*, cli.estadocivil FROM `guilherme-454214.locadora.aluguel` a JOIN `guilherme-454214.locadora.cliente` cli ON a.codcliente = cli.codcliente

✅ DataFrame 'Q33' criado com sucesso! (10 linhas)


codaluguel  codcliente  codcarro data_aluguel estadocivil
0           2           2         1   2023-04-02           C
1           3           2         1   2023-04-03           C
2           7           1         1   2023-04-15           C
3          10           3         1   2023-04-25           S
4           1           3         2   2023-04-01           S
5           8           5         2   2023-04-19           C
6           9           5         2   2023-04-21           C
7           4           2         3   2023-04-04           C
8           5           1         4   2023-04-05           C
9           6           1         4   2023-04-13           C

## Exercício 34

Liste os clientes e as cidades em que moram para aqueles que alugaram carros.

```sql
-- Escreva sua consulta aqui
```

In [57]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT DISTINCT cli.nome, cli.cidade FROM cliente cli JOIN aluguel a ON cli.codcliente = a.codcliente
Informe o nome para o DataFrame:
Q34
SELECT DISTINCT cli.nome, cli.cidade FROM `guilherme-454214.locadora.cliente` cli JOIN `guilherme-454214.locadora.aluguel` a ON cli.codcliente = a.codcliente

✅ DataFrame 'Q34' criado com sucesso! (4 linhas)


nome           cidade
0         Ana Silva  Duque de Caxias
1     Bruna Pereira          Niterói
2  Túlio Nascimento  Duque de Caxias
3     Lúcia Andrade        São Paulo

## Exercício 35

Liste os clientes que alugaram carros e os modelos de carros que eles alugaram, incluindo a marca do carro.

```sql
-- Escreva sua consulta aqui
```

In [58]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT cli.nome, ca.modelo, ma.marca FROM cliente cli JOIN aluguel a ON cli.codcliente = a.codcliente JOIN carro ca ON a.codcarro = ca.codcarro JOIN marca ma ON ca.codmarca = ma.codmarca
Informe o nome para o DataFrame:
Q35
SELECT cli.nome, ca.modelo, ma.marca FROM `guilherme-454214.locadora.cliente` cli JOIN `guilherme-454214.locadora.aluguel` a ON cli.codcliente = a.codcliente JOIN `guilherme-454214.locadora.carro` ca ON a.codcarro = ca.codcarro JOIN `guilherme-454214.locadora.marca` ma ON ca.codmarca = ma.codmarca

✅ DataFrame 'Q35' criado com sucesso! (10 linhas)


nome modelo       marca
0         Ana Silva     Ka        Ford
1         Ana Silva   Polo  Volkswagen
2         Ana Silva   Polo  Volkswagen
3     Bruna Pereira     Ka        Ford
4     Bruna Pereira     Ka        Ford
5     Bruna Pereira   Onix   Chevrolet
6  Túlio Nascimento     Ka        Ford
7  Túlio Nascimento   Argo        Fiat
8     Lúcia Andrade   Argo        Fiat
9     Lúcia Andrade   Argo        Fiat

## Exercício 36

Encontre todos os aluguéis realizados em uma cidade específica e liste os nomes dos clientes e os modelos de carros alugados.

```sql
-- Escreva sua consulta aqui
```

In [61]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT cli.nome, ca.modelo FROM cliente cli JOIN aluguel a ON cli.codcliente = a.codcliente JOIN carro ca ON a.codcarro = ca.codcarro WHERE cli.cidade = 'Duque de Caxias'
Informe o nome para o DataFrame:
Q36
SELECT cli.nome, ca.modelo FROM `guilherme-454214.locadora.cliente` cli JOIN `guilherme-454214.locadora.aluguel` a ON cli.codcliente = a.codcliente JOIN `guilherme-454214.locadora.carro` ca ON a.codcarro = ca.codcarro WHERE cli.cidade = 'Duque de Caxias'

✅ DataFrame 'Q36' criado com sucesso! (5 linhas)


nome modelo
0         Ana Silva     Ka
1         Ana Silva   Polo
2         Ana Silva   Polo
3  Túlio Nascimento     Ka
4  Túlio Nascimento   Argo

## Exercício 37

Liste os modelos de carros e a quantidade de vezes que foram alugados.

```sql
-- Escreva sua consulta aqui
```

In [62]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT ca.modelo, COUNT(a.codaluguel) AS total_alugueis FROM carro ca LEFT JOIN aluguel a ON ca.codcarro = a.codcarro GROUP BY ca.modelo
Informe o nome para o DataFrame:
Q37
SELECT ca.modelo, COUNT(a.codaluguel) AS total_alugueis FROM `guilherme-454214.locadora.carro` ca LEFT JOIN `guilherme-454214.locadora.aluguel` a ON ca.codcarro = a.codcarro GROUP BY ca.modelo

✅ DataFrame 'Q37' criado com sucesso! (5 linhas)


modelo  total_alugueis
0     Ka               4
1   Kwid               0
2   Argo               3
3   Polo               2
4   Onix               1

## Exercício 38

Encontre os clientes que alugaram mais de um carro.

```sql
-- Escreva sua consulta aqui
```

In [63]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT codcliente, COUNT(*) AS total_alugueis FROM aluguel GROUP BY codcliente HAVING total_alugueis > 1 
Informe o nome para o DataFrame:
Q38
SELECT codcliente, COUNT(*) AS total_alugueis FROM `guilherme-454214.locadora.aluguel` GROUP BY codcliente HAVING total_alugueis > 1 

✅ DataFrame 'Q38' criado com sucesso! (4 linhas)


codcliente  total_alugueis
0           2               3
1           1               3
2           3               2
3           5               2

## Exercício 39

Liste os clientes, as datas de aluguel e o valor do carro alugado.

```sql
-- Escreva sua consulta aqui
```

In [64]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT cli.nome, a.data_aluguel, ca.valor FROM cliente cli JOIN aluguel a ON cli.codcliente = a.codcliente JOIN carro ca ON a.codcarro = ca.codcarro
Informe o nome para o DataFrame:
Q39
SELECT cli.nome, a.data_aluguel, ca.valor FROM `guilherme-454214.locadora.cliente` cli JOIN `guilherme-454214.locadora.aluguel` a ON cli.codcliente = a.codcliente JOIN `guilherme-454214.locadora.carro` ca ON a.codcarro = ca.codcarro

✅ DataFrame 'Q39' criado com sucesso! (10 linhas)


nome data_aluguel  valor
0         Ana Silva   2023-04-15    100
1         Ana Silva   2023-04-05    150
2         Ana Silva   2023-04-13    150
3     Bruna Pereira   2023-04-02    100
4     Bruna Pereira   2023-04-03    100
5     Bruna Pereira   2023-04-04    170
6  Túlio Nascimento   2023-04-25    100
7  Túlio Nascimento   2023-04-01    150
8     Lúcia Andrade   2023-04-19    150
9     Lúcia Andrade   2023-04-21    150

## Exercício 40

Liste os modelos de carros que nunca foram alugados.

```sql
-- Escreva sua consulta aqui
```

In [65]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT modelo FROM carro WHERE codcarro NOT IN (SELECT codcarro FROM aluguel)
Informe o nome para o DataFrame:
Q40
SELECT modelo FROM `guilherme-454214.locadora.carro` WHERE codcarro NOT IN (SELECT codcarro FROM `guilherme-454214.locadora.aluguel`)

✅ DataFrame 'Q40' criado com sucesso! (1 linhas)


modelo
0   Kwid

## Exercício 41

Encontre os clientes que alugaram carros e o total gasto em aluguéis.

```sql
-- Escreva sua consulta aqui
```

In [70]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT cli.nome, SUM(ca.valor) AS total_gasto  FROM cliente cli JOIN aluguel a ON cli.codcliente = a.codcliente JOIN carro ca ON a.codcarro = ca.codcarro GROUP BY cli.codcliente, cli.nome
Informe o nome para o DataFrame:
Q41
SELECT cli.nome, SUM(ca.valor) AS total_gasto  FROM `guilherme-454214.locadora.cliente` cli JOIN `guilherme-454214.locadora.aluguel` a ON cli.codcliente = a.codcliente JOIN `guilherme-454214.locadora.carro` ca ON a.codcarro = ca.codcarro GROUP BY cli.codcliente, cli.nome

✅ DataFrame 'Q41' criado com sucesso! (4 linhas)


nome  total_gasto
0         Ana Silva          400
1     Bruna Pereira          370
2  Túlio Nascimento          250
3     Lúcia Andrade          300

## Exercício 42

Liste as marcas de carros e o número total de aluguéis por marca.

```sql
-- Escreva sua consulta aqui
```

In [68]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT ma.marca, COUNT(a.codaluguel) AS total_alugueis FROM marca ma LEFT JOIN carro ca ON ma.codmarca = ca.codmarca LEFT JOIN aluguel a ON ca.codcarro = a.codcarro GROUP BY ma.marca
Informe o nome para o DataFrame:
Q42
SELECT ma.marca, COUNT(a.codaluguel) AS total_alugueis FROM `guilherme-454214.locadora.marca` ma LEFT JOIN `guilherme-454214.locadora.carro` ca ON ma.codmarca = ca.codmarca LEFT JOIN `guilherme-454214.locadora.aluguel` a ON ca.codcarro = a.codcarro GROUP BY ma.marca

✅ DataFrame 'Q42' criado com sucesso! (5 linhas)


marca  total_alugueis
0        Ford               4
1        Fiat               3
2   Chevrolet               1
3  Volkswagen               2
4     Renault               0

## Exercício 43

Encontre os clientes que alugaram carros em um determinado mês e ano, listando os modelos de carros e as datas de aluguel.

```sql
-- Escreva sua consulta aqui
```

In [99]:
consulta()

KeyboardInterrupt: Interrupted by user

## Exercício 44

Liste os clientes que moram em estados diferentes e que alugaram carros, junto com os modelos alugados.

```sql
-- Escreva sua consulta aqui
```

In [74]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT c1.nome AS cliente1, c1.estado AS estado1, c2.nome AS cliente2, c2.estado AS estado2, ca.modelo FROM aluguel a1 JOIN aluguel a2 ON a1.codcarro = a2.codcarro AND a1.codcliente < a2.codcliente JOIN cliente c1 ON a1.codcliente = c1.codcliente JOIN cliente c2 ON a2.codcliente = c2.codcliente JOIN carro ca ON a1.codcarro = ca.codcarro WHERE c1.estado <> c2.estado
Informe o nome para o DataFrame:
Q44
SELECT c1.nome AS cliente1, c1.estado AS estado1, c2.nome AS cliente2, c2.estado AS estado2, ca.modelo FROM `guilherme-454214.locadora.aluguel` a1 JOIN `guilherme-454214.locadora.aluguel` a2 ON a1.codcarro = a2.codcarro AND a1.codcliente < a2.codcliente JOIN `guilherme-454214.locadora.cliente` c1 ON a1.codcliente = c1.codcliente JOIN `guilherme-454214.locadora.cliente` c2 ON a2.codcliente = c2.codcliente JOIN `guilherme-454214.locadora.carro` ca ON a1.codcarro = ca.codcarro WHERE c1.estado <> c2.estado

✅ DataFrame 'Q44' criado com sucesso! (2 li

cliente1 estado1       cliente2 estado2 modelo
0  Túlio Nascimento      RJ  Lúcia Andrade      SP   Argo
1  Túlio Nascimento      RJ  Lúcia Andrade      SP   Argo

## Exercício 45

Encontre os clientes que alugaram carros, os modelos de carros alugados, as marcas desses carros e as datas de aluguel, ordenando pelos nomes dos clientes.

```sql
-- Escreva sua consulta aqui
```

In [77]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT cli.nome, ca.modelo, ma.marca, a.data_aluguel FROM cliente cli JOIN aluguel a ON cli.codcliente = a.codcliente JOIN carro ca ON a.codcarro = ca.codcarro JOIN marca ma ON ca.codmarca = ma.codmarca ORDER BY cli.nome
Informe o nome para o DataFrame:
Q45
SELECT cli.nome, ca.modelo, ma.marca, a.data_aluguel FROM `guilherme-454214.locadora.cliente` cli JOIN `guilherme-454214.locadora.aluguel` a ON cli.codcliente = a.codcliente JOIN `guilherme-454214.locadora.carro` ca ON a.codcarro = ca.codcarro JOIN `guilherme-454214.locadora.marca` ma ON ca.codmarca = ma.codmarca ORDER BY cli.nome

✅ DataFrame 'Q45' criado com sucesso! (10 linhas)


nome modelo       marca data_aluguel
0         Ana Silva     Ka        Ford   2023-04-15
1         Ana Silva   Polo  Volkswagen   2023-04-05
2         Ana Silva   Polo  Volkswagen   2023-04-13
3     Bruna Pereira     Ka        Ford   2023-04-02
4     Bruna Pereira     Ka        Ford   2023-04-03
5     Bruna Pereira   Onix   Chevrolet   2023-04-04
6     Lúcia Andrade   Argo        Fiat   2023-04-19
7     Lúcia Andrade   Argo        Fiat   2023-04-21
8  Túlio Nascimento     Ka        Ford   2023-04-25
9  Túlio Nascimento   Argo        Fiat   2023-04-01

## Exercício 46

Liste os clientes que alugaram carros mais de uma vez e os detalhes dos carros que alugaram.

```sql
-- Escreva sua consulta aqui
```

In [78]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT cli.nome, ca.modelo, ma.marca, a.data_aluguel FROM cliente cli JOIN aluguel a ON cli.codcliente = a.codcliente JOIN carro ca ON a.codcarro = ca.codcarro JOIN marca ma ON ca.codmarca = ma.codmarca WHERE cli.codcliente IN ( SELECT codcliente FROM aluguel GROUP BY codcliente HAVING COUNT(*) > 1 )
Informe o nome para o DataFrame:
Q46
SELECT cli.nome, ca.modelo, ma.marca, a.data_aluguel FROM `guilherme-454214.locadora.cliente` cli JOIN `guilherme-454214.locadora.aluguel` a ON cli.codcliente = a.codcliente JOIN `guilherme-454214.locadora.carro` ca ON a.codcarro = ca.codcarro JOIN `guilherme-454214.locadora.marca` ma ON ca.codmarca = ma.codmarca WHERE cli.codcliente IN ( SELECT codcliente FROM `guilherme-454214.locadora.aluguel` GROUP BY codcliente HAVING COUNT(*) > 1 )

✅ DataFrame 'Q46' criado com sucesso! (10 linhas)


nome modelo       marca data_aluguel
0         Ana Silva     Ka        Ford   2023-04-15
1         Ana Silva   Polo  Volkswagen   2023-04-05
2         Ana Silva   Polo  Volkswagen   2023-04-13
3     Bruna Pereira     Ka        Ford   2023-04-02
4     Bruna Pereira     Ka        Ford   2023-04-03
5     Bruna Pereira   Onix   Chevrolet   2023-04-04
6  Túlio Nascimento     Ka        Ford   2023-04-25
7  Túlio Nascimento   Argo        Fiat   2023-04-01
8     Lúcia Andrade   Argo        Fiat   2023-04-19
9     Lúcia Andrade   Argo        Fiat   2023-04-21

## Exercício 47

Encontre os modelos de carros que foram alugados por clientes de cidades específicas e liste as marcas desses carros.

```sql
-- Escreva sua consulta aqui
```

In [82]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT ca.modelo, ma.marca  FROM carro ca JOIN aluguel a ON ca.codcarro = a.codcarro JOIN marca ma ON ca.codmarca = ma.codmarca JOIN cliente cli ON a.codcliente = cli.codcliente WHERE cli.cidade IN ('Duque de Caxias', 'São Paulo');
Informe o nome para o DataFrame:
Q47
SELECT ca.modelo, ma.marca  FROM `guilherme-454214.locadora.carro` ca JOIN `guilherme-454214.locadora.aluguel` a ON ca.codcarro = a.codcarro JOIN `guilherme-454214.locadora.marca` ma ON ca.codmarca = ma.codmarca JOIN `guilherme-454214.locadora.cliente` cli ON a.codcliente = cli.codcliente WHERE cli.cidade IN ('Duque de Caxias', 'São Paulo');

✅ DataFrame 'Q47' criado com sucesso! (7 linhas)


modelo       marca
0     Ka        Ford
1     Ka        Ford
2   Argo        Fiat
3   Argo        Fiat
4   Argo        Fiat
5   Polo  Volkswagen
6   Polo  Volkswagen

## Exercício 48

Liste os clientes que alugaram carros, os modelos e as marcas dos carros, e o total gasto em aluguéis, agrupado por cliente.

```sql
-- Escreva sua consulta aqui
```

In [89]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT cli.nome, ca.modelo, ma.marca, SUM(ca.valor) AS total_gasto  FROM cliente cli JOIN aluguel a ON cli.codcliente = a.codcliente JOIN carro ca ON a.codcarro = ca.codcarro JOIN marca ma ON ca.codmarca = ma.codmarca GROUP BY cli.nome, ca.modelo, ma.marca, cli.codcliente, ca.codcarro
Informe o nome para o DataFrame:
Q48
SELECT cli.nome, ca.modelo, ma.marca, SUM(ca.valor) AS total_gasto  FROM `guilherme-454214.locadora.cliente` cli JOIN `guilherme-454214.locadora.aluguel` a ON cli.codcliente = a.codcliente JOIN `guilherme-454214.locadora.carro` ca ON a.codcarro = ca.codcarro JOIN `guilherme-454214.locadora.marca` ma ON ca.codmarca = ma.codmarca GROUP BY cli.nome, ca.modelo, ma.marca, cli.codcliente, ca.codcarro

✅ DataFrame 'Q48' criado com sucesso! (7 linhas)


nome modelo       marca  total_gasto
0         Ana Silva     Ka        Ford          100
1         Ana Silva   Polo  Volkswagen          300
2     Bruna Pereira     Ka        Ford          200
3     Bruna Pereira   Onix   Chevrolet          170
4  Túlio Nascimento     Ka        Ford          100
5  Túlio Nascimento   Argo        Fiat          150
6     Lúcia Andrade   Argo        Fiat          300

## Exercício 49

Encontre os carros que foram alugados nos últimos 30 dias e os clientes que os alugaram.

```sql
-- Escreva sua consulta aqui
```

In [94]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT cli.nome AS cliente, ca.modelo AS carro, a.data_aluguel FROM `guilherme-454214.locadora.cliente` cli JOIN `guilherme-454214.locadora.aluguel` a    ON cli.codcliente = a.codcliente JOIN `guilherme-454214.locadora.carro` ca    ON a.codcarro = ca.codcarro WHERE a.data_aluguel >= DATE_SUB(CURRENT_DATE(), INTERVAL 30 DAY) ORDER BY a.data_aluguel DESC;
Informe o nome para o DataFrame:
Q49
SELECT cli.nome AS cliente, ca.modelo AS carro, a.data_aluguel FROM `guilherme-454214.locadora.cliente` cli JOIN `guilherme-454214.locadora.aluguel` a    ON cli.codcliente = a.codcliente JOIN `guilherme-454214.locadora.carro` ca    ON a.codcarro = ca.codcarro WHERE a.data_aluguel >= DATE_SUB(CURRENT_DATE(), INTERVAL 30 DAY) ORDER BY a.data_aluguel DESC;

✅ DataFrame 'Q49' criado com sucesso! (0 linhas)


Empty DataFrame
Columns: [cliente, carro, data_aluguel]
Index: []

## Exercício 50

Liste os clientes, os carros que alugaram, as marcas desses carros, e o valor total gasto em aluguéis, filtrando por um intervalo de datas específico.

```sql
-- Escreva sua consulta aqui
```

In [93]:
consulta()

Digite sua query (sem project_id.dataset):
SELECT cli.nome, ca.modelo, ma.marca, SUM(ca.valor) AS total_gasto  FROM cliente cli JOIN aluguel a ON cli.codcliente = a.codcliente JOIN carro ca ON a.codcarro = ca.codcarro JOIN marca ma ON ca.codmarca = ma.codmarca WHERE a.data_aluguel BETWEEN '2023-04-01' AND '2023-04-30' GROUP BY cli.codcliente, ca.codcarro, cli.nome, ca.modelo, ma.marca
Informe o nome para o DataFrame:
Q50
SELECT cli.nome, ca.modelo, ma.marca, SUM(ca.valor) AS total_gasto  FROM `guilherme-454214.locadora.cliente` cli JOIN `guilherme-454214.locadora.aluguel` a ON cli.codcliente = a.codcliente JOIN `guilherme-454214.locadora.carro` ca ON a.codcarro = ca.codcarro JOIN `guilherme-454214.locadora.marca` ma ON ca.codmarca = ma.codmarca WHERE a.data_aluguel BETWEEN '2023-04-01' AND '2023-04-30' GROUP BY cli.codcliente, ca.codcarro, cli.nome, ca.modelo, ma.marca

✅ DataFrame 'Q50' criado com sucesso! (7 linhas)


nome modelo       marca  total_gasto
0         Ana Silva     Ka        Ford          100
1         Ana Silva   Polo  Volkswagen          300
2     Bruna Pereira     Ka        Ford          200
3     Bruna Pereira   Onix   Chevrolet          170
4  Túlio Nascimento     Ka        Ford          100
5  Túlio Nascimento   Argo        Fiat          150
6     Lúcia Andrade   Argo        Fiat          300